pkg/asset/fsck.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package asset
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+)
+
+// detectFileSystemType uses blkid to identify the file system type of a given image file
+func detectFileSystemType(imagePath string) (string, error) {
+	cmd := exec.Command("blkid", "-o", "value", "-s", "TYPE", imagePath)
+	output, err := cmd.Output()
+	if err != nil {
+		return "", fmt.Errorf("failed to detect file system type: %v", err)
+	}
+	return string(output), nil
+}
+
+// runFsck uses the fs-specific fsck subcommand to check if an image file is clean
+func runFsck(imagePath string, fsType string) (bool, error) {
+	fsckCommand := fmt.Sprintf("fsck.%s", fsType)
+	cmd := exec.Command(fsckCommand, "-n", imagePath)
```

Comment:
I don't see -n flag in "man fsck.btrfs". Do all of these support the flag?

Reply:
How big this map would be?

If the follow the map approach, we could determine the filesystem name directly from the system call from which the fs image was extracted.

Reply:
> How big this map would be?

`ls sys/linux/test/syz_mount_image_*_0 | wc -l` gives me `41` so I guess the map would ideally get that big but I also don't think all these filesystems necessarily come with a fsck.

As far as I can tell there are 26 `fsck.*` variants in the debian repositories
https://packages.debian.org/search?searchon=contents&keywords=fsck&mode=filename&suite=stable&arch=any but some of them for fs we don't fuzz (eg: winregfs)

I'm wondering whether we should make that map part of syz-manager config so that users can configure the checks they want to run... Then we don't have to make assumptions about what commands are available. Or we can hardcode the map in golang but we can let the user provide the list of fstypes they want to check. Or we dynamically try to resolve which commands are available and have a clean fallback for missing commands or so.

> If the follow the map approach, we could determine the filesystem name directly from the system call from which the fs image was extracted.

I guess we'd have to refactor the signature of callbacks provided to `ForEachAsset()` a little bit but I suppose we can find a way to make it work yeah

Reply:
> I'm wondering whether we should make that map part of syz-manager config so that users can configure the checks they want to run... Then we don't have to make assumptions about what commands are available. Or we can hardcode the map in golang but we can let the user provide the list of fstypes they want to check. Or we dynamically try to resolve which commands are available and have a clean fallback for missing commands or so.

For local runs, AFAIK we don't extract mounted images at all, so I'd say that it's unlikely that external users will notice or care.

If we want to avoid hardcoding syscall names into the code, we could theoretically consider setting the proper mkfs binary name directly as a syscall attribute here

https://github.com/google/syzkaller/blob/52b38cc1f0ef7be017dc51c45e16a1054254f95f/prog/types.go#L39

> I guess we'd have to refactor the signature of callbacks provided to ForEachAsset() a little bit but I suppose we can find a way to make it work yeah

Sure, we can change them however we need to :)